### ISSUE TITLE
spy.getCallAsync(n)

### ISSUE THREAD

**Issue by Havvy** [cjohansen/Sinon.JS#731]
Running in Node.js & io.js here, I've got some asyncronous code where I want to check that the third call to a function passed specific arguments^. Unfortunately, I don't have any hooks that I can use to check when the function is called, so I can't wait on a hook to fire. Instead, I'd have to modify the spy function to fire off an event when called or something. But really, I'd just be adding the feature I'm proposing in this issue in an ad hoc manner.

Basically, add a method to spies called "getCallAsync" or something similar that returns a Promise<SpyCall, null>. The reason the Error side is `null` is because the current spy.getCall(n) doesn't throw an error. Usage would be something like this:

```javascript
// Using `mocha` as the testing environment.
var sinon = require("sinon");

it("gets a sinon spy call asyncronously", function () {
    var spy = sinon.spy();

    var promise = spy.getCallAsync(1)
    .then(function (spyCall) {
        assert (spyCall.calledWithExactly("foo", "bar");
    });

    setImmediate(function () {
        spy("foo", "bar");
    });

    return promise;
});
```

There are other methods that could have variants that return promises, but this one would be most helpful to me, and be a good experimenting ground before adding others.

If you don't want to depend on a packaged Promise library, I'd be okay with injecting one in myself via some method such as `sinon.usePromiseLibrary(require("bluebird"));` Perhaps Sinon could detect the global "Promise" and default to using it if it exists.

^ If that sounds brittle, it sort of is, but it'd be brittle synchronously too. Something less brittle would be waiting for the function to be called with specific arguments.

**Comment by mantoni** [comment]
Couldn't you do this with a custom spy function?

```js
var spy = sinon.spy(function () {
  // any custom logic on function call
});
```

**Comment by Havvy** [comment]
Yes and no. If I'm creating the spy in the same scope as the test, it'd be trivial.

```javascript
it("test", function () {
    var spy = sinon.spy(function (...args) {
        if (spy.callCount === 3) {
             assert(spy.getCall(3).calledWithExactly(...);
        }
    };

    ....;
});
```

But I'm actually creating the spy outside of that scope in a function that I'm using for all of those tests and it would be non-trivial or highly unreadable to modify the spy in question.

Ultimately, for my purposes, I monkeypatched a .on(n, fn) & .emit() function onto the spy and called .emit() on the spy. But that's not clean at all, and I had to leave a comment explaining why I did that.

**Comment by mantoni** [comment]
Hm. Maybe you can post a description of your problem on the mailing list (instead of suggesting a solution beforehand). There might be other ways to improve your tests and the audience on the mailing list is a bit lerger than on the issue tracker.

**Comment by mroderick** [comment]
We are trying to keep the GitHub issues list tidy and focused on bugs and feature discussions. This ticket looks like a usage question, please post it to the [Sinon.JS mailinglist](http://groups.google.com/group/sinonjs), so the bigger community can help answer your questions.

If you feel that your topic is an issue with Sinon.JS, please open a new ticket and follow the [guidelines for reporting an issue](https://github.com/cjohansen/Sinon.JS/blob/master/CONTRIBUTING.md#reporting-an-issue).

**Comment by felixfbecker** [comment]
Would really like this feature too.
